02_classify_stratified/run_purity_prediction.py

```diff
@@ -0,0 +1,211 @@
+                              n_dim=model_options.n_dim,
+                              sample_info_df=sample_info_df,
+                              verbose=io_args.verbose,
+                              debug=model_options.debug)
+
+    # we want to run purity prediction experiments for true labels and
+    # shuffled labels (the latter as a lower baseline)
+    progress = tqdm([False, True],
+                    ncols=100,
+                    file=sys.stdout)
+    for shuffle_labels in progress:
+        progress.set_description('shuffle labels: {}'.format(shuffle_labels))
+
+        try:
+            output_dir = fu.make_output_dir(experiment_dir, '')
+            check_file = fu.check_output_file(output_dir,
+                                              None,
+                                              shuffle_labels,
+                                              model_options)
+        except ResultsFileExistsError:
+            # this happens if cross-validation for this gene has already been
+            # run (i.e. the results file already exists)
+            if io_args.verbose:
+                print('Skipping because results file exists already', file=sys.stderr)
+            purity_log_df = fu.generate_log_df(
+                log_columns,
+                [model_options.training_data, shuffle_labels, 'file_exists']
+            )
+            fu.write_log_file(purity_log_df, io_args.log_file)
+            continue
+
+        tcga_data.process_purity_data(experiment_dir,
+                                      shuffle_labels=shuffle_labels)
+
+        try:
+            # for now, don't standardize methylation data
+            standardize_columns = (model_options.training_data in
+                                   cfg.standardize_data_types)
+            results = run_cv_stratified(tcga_data,
+                                        'purity',
+                                        None,
+                                        model_options.training_data,
+                                        sample_info_df,
+                                        model_options.num_folds,
+                                        shuffle_labels,
+                                        standardize_columns,
+                                        io_args.output_preds)
+        except NoTrainSamplesError:
+            if io_args.verbose:
+                print('Skipping due to no train samples', file=sys.stderr)
+            purity_log_df = fu.generate_log_df(
+                log_columns,
+                [model_options.training_data, shuffle_labels, 'no_train_samples']
+            )
+        except OneClassError:
+            if io_args.verbose:
+                print('Skipping due to one holdout class', file=sys.stderr)
+            purity_log_df = fu.generate_log_df(
+                log_columns,
+                [model_options.training_data, shuffle_labels, 'one_class']
+            )
+        else:
```

Comment:
Looking at [this post](https://stackoverflow.com/a/855764), I think you're right. Since I'm only catching exceptions that I defined (and that won't happen in the `save_results` function), there shouldn't be any difference here between adding the `else` clause and putting the save code on line 183.

I'll switch to what you're suggesting, since I think it'll be clearer to most people.